File: 3. Behavioral Design Patterns/Memento/Go/main.go
package main

import "fmt"

// Memento Pattern Example in Go
// Text editor that can save and restore document states

type TextMemento struct {
	content string
}

func NewTextMemento(content string) *TextMemento {
	return &TextMemento{content: content}
}

func (m *TextMemento) GetContent() string {
	return m.content
}

type TextEditor struct {
	content string
}

func NewTextEditor() *TextEditor {
	return &TextEditor{content: ""}
}

func (e *TextEditor) Write(text string) {
	e.content += text
}

func (e *TextEditor) Clear() {
	e.content = ""
}

func (e *TextEditor) GetContent() string {
	return e.content
}

func (e *TextEditor) Save() *TextMemento {
	return NewTextMemento(e.content)
}

func (e *TextEditor) Restore(memento *TextMemento) {
	e.content = memento.GetContent()
}

type EditorHistory struct {
	history []*TextMemento
}

func NewEditorHistory() *EditorHistory {
	return &EditorHistory{history: make([]*TextMemento, 0)}
}

func (h *EditorHistory) Save(editor *TextEditor) {
	h.history = append(h.history, editor.Save())
}

func (h *EditorHistory) Undo(editor *TextEditor) {
	if len(h.history) > 0 {
		memento := h.history[len(h.history)-1]
		h.history = h.history[:len(h.history)-1]
		editor.Restore(memento)
	}
}

func main() {
	editor := NewTextEditor()
	history := NewEditorHistory()

	editor.Write("Hello ")
	history.Save(editor)
	fmt.Println("Content:", editor.GetContent())

	editor.Write("World!")
	history.Save(editor)
	fmt.Println("Content:", editor.GetContent())

	history.Undo(editor)
	fmt.Println("After undo:", editor.GetContent())

	history.Undo(editor)
	fmt.Println("After undo:", editor.GetContent())
}

File: 3. Behavioral Design Patterns/Memento/Rust/main.rs
// Memento Pattern Example in Rust
// Text editor that can save and restore document states

#[derive(Clone)]
struct TextMemento {
    content: String,
}

impl TextMemento {
    fn new(content: String) -> Self {
        Self { content }
    }

    fn get_content(&self) -> &str {
        &self.content
    }
}

struct TextEditor {
    content: String,
}

impl TextEditor {
    fn new() -> Self {
        Self {
            content: String::new(),
        }
    }

    fn write(&mut self, text: &str) {
        self.content.push_str(text);
    }

    fn clear(&mut self) {
        self.content.clear();
    }

    fn get_content(&self) -> &str {
        &self.content
    }

    fn save(&self) -> TextMemento {
        TextMemento::new(self.content.clone())
    }

    fn restore(&mut self, memento: &TextMemento) {
        self.content = memento.get_content().to_string();
    }
}

struct EditorHistory {
    history: Vec<TextMemento>,
}

impl EditorHistory {
    fn new() -> Self {
        Self {
            history: Vec::new(),
        }
    }

    fn save(&mut self, editor: &TextEditor) {
        self.history.push(editor.save());
    }

    fn undo(&mut self, editor: &mut TextEditor) {
        if let Some(memento) = self.history.pop() {
            editor.restore(&memento);
        }
    }
}

fn main() {
    let mut editor = TextEditor::new();
    let mut history = EditorHistory::new();

    editor.write("Hello ");
    history.save(&editor);
    println!("Content: {}", editor.get_content());

    editor.write("World!");
    history.save(&editor);
    println!("Content: {}", editor.get_content());

    history.undo(&mut editor);
    println!("After undo: {}", editor.get_content());

    history.undo(&mut editor);
    println!("After undo: {}", editor.get_content());
}
